keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1014 @@
+were alpha in the Kubernetes version the compatibility version is set to. One
+important implication of this requirement is that feature gating must be kept in
+the Kubenetes codebase until a feature has reached GA (or been removed) for N-3
+releases.
+
+For example, a feature that is promoted once per release would look something like:
+
+```go
+map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.27"), Default: true, PreRelease: Beta},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: GA},
+		},
+}
+```
+
+The lifecycle of the feature would be:
+
+| Release | Stage | Feature tracking information                      |
+| ------- | ----- | ------------------------------------------------- |
+| 1.26    | alpha | Alpha: 1.26                                       |
+| 1.27    | beta  | Alpha: 1.26, Beta: 1.27 (on-by-default)           |
+| 1.28    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.29    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.30    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.31    | GA    | **Feature implementation becomes part of normal code, Feature gate and feature tracking information may be removed from code** |
+
+All feature gating and tracking must remain in code through 1.30 for N-3
+compatibility version support.
+
+For a feature that is removed, e.g.:
+
+```go
+map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.27"), Default: false, PreRelease: Deprecated},
+			{Version: mustParseVersion("1.31"), Default: false, PreRelease: Removed},
+		},
+}
+```
+
+The lifecycle of the feature implementation would be:
+
+| Release | Stage | Feature tracking information                      |
+| ------- | ----- | ------------------------------------------------- |
+| 1.26    | alpha | Alpha: 1.26                                       |
+| 1.27    | alpha | Alpha: 1.26, Deprecated: 1.27                     |
+| 1.28    | alpha | Alpha: 1.26, Deprecated: 1.27                     |
+| 1.29    | alpha | Alpha: 1.26, Deprecated: 1.27                     |
+| 1.30    | -     | Alpha: 1.26, Deprecated: 1.27, Removed: 1.31      |
+| 1.31    | -     | Alpha: 1.26, Deprecated: 1.27, Removed: 1.31      |
+| 1.32    | -     | Alpha: 1.26, Deprecated: 1.27, Removed: 1.31      |
+| 1.33    | -     | **Feature implementation and feature tracking information may be removed from code** |
+
+Note that this respects a 1yr deprecation policy.
+
+All feature gating and tracking must remain in code through 1.32 for N-3
+compatibility version support.
+
+#### Feature gating changes
+
+In order to preserve the behavior of in-development features across multiple releases,
+feature implementations may also be gated by version number.
```

Comment:
Is this intended to read as MAY or MUST?

While I understand the spirit behind the option, it seems like it may complicate any automated cherry pick process.  Is this intended to maintain bug for bug compatibility across binary versions, or is this intended to be less about bug for bug compatibility and more evolution of feature capability?

Reply:
I agree with the concern. We don't want to make the unreasonably burdensome to feature authors.

We don't want feature authors to retain bug for bug compatibility.  The intent here is to preserve significant and observable differences in feature capabilities across releases.  

@siyuanfoundation just added some suggestions on how to make handle this mechanically in a cleaner and less error prone way: https://github.com/kubernetes/enhancements/pull/4395/commits/d2a0caa2178aab929290d2afef7c64ac6882b693 

I'll update the KEP with some wording that clarifies our expectations.